File: hackerrank/Cut the Rope.cpp
#include<iostream>
#include<cstdio>
#include<cmath>
using namespace std;


int T, n;
long long dp[100500], ans[100500];

double getLog(int x){
	return log(x)/log(2.0);
} 

long long deal(int x){
	if (dp[x] > 0) return dp[x];
	for (long long i = 2; i <= x/2; i++){
		long long x1, y1, y2, x2, j = x-i;
		x1 = deal(i); x1 = max(x1, i);
		x2 = deal(j); x2 = max(x2, j);
		dp[x] = max(dp[x], x1*x2);
	}
	return dp[x];
}

int main(){
	cin >> T;
	dp[2] = 1;
	dp[3] = 2;
	for (int i = 0; i < T; i++){
		cin >> n;
		cout << "output:" << deal(n) << endl;
	}
	
}

File: hackerrank/Cut the Rope.py
T = int(input())
dp = [0] * 100500
dp[2] = 1
dp[3] = 2

def deal(x):
    if dp[x] > 0: return dp[x]
    for i in range(2, x//2+1):
        dp[x] = max(dp[x], max(deal(i), i) * max(deal(x-i), x-i))
    return dp[x]

while T > 0:
    T -= 1
    n = int(input())
    # print("output: "+str(deal(n) % 1000))
    print(deal(n)%1000)
    
